### PR TITLE
Fix SGD and Adafactor weight decay mutating caller arrays

### PR DESCRIPTION
**Pull request by 2sumtech**
- ☑️ I understand it is strictly prohibited to use AI to write PR description
changed a += b to a = a + b. so a stays as a, not confuse training.

### CODE REVIEW

**Comment by nastya236** [comment]
Could you specify what bug you observed?

**Comment by 2sumtech** [comment]
```python
import numpy as np
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim

model = nn.Linear(3, 2)
opt = optim.SGD(learning_rate=0.1, weight_decay=0.5)
grads = {"weight": mx.ones((2, 3)), "bias": mx.ones((2,))}
opt.update(model, grads)
mx.eval(model.parameters(), grads)
print(np.array(grads["weight"])[0])   # expected [1. 1. 1.], actual values are changed
```
when running like this

**Comment by nastya236** [comment]
Thank you, it might be useful if one wants to inspect gradients with SGD.

**Review by nastya236 — APPROVED** [review]
